### PR TITLE
TLS and external endpoint support for shared infra

### PR DESCRIPTION
**Pull request by lulf**
* Add support for configuring TLS endpoints.
* Order creation of SslProfile and Listener.
* Add support for Route and Ingress endpoints.
* Add support for websockets.
* Add support for selfsigned, openshift and external certificates.
* Add systemtests for different combinations of features.

### CODE REVIEW

**Comment by lulf** [comment]
@enmasse-ci run tests profile=isolated-shared-infra

**Comment by lulf** [comment]
@enmasse-ci run tests profile=isolated=shared-infra

**Comment by lulf** [comment]
@enmasse-ci run tests profile=isolated-shared-infra

**Comment by lulf** [comment]
@enmasse-ci run tests profile=isolated-shared-infra

**Comment by lulf** [comment]
@enmasse-ci run tests profile=isolated-shared-infra

**Comment by lulf** [comment]
@enmasse-ci run tests profile=isolated-shared-infra

**Comment by lulf** [comment]
@enmasse-ci run tests profile=isolated-shared-infra

**Comment by lulf** [comment]
@enmasse-ci run tests profile=isolated-shared-infra

**Comment by lulf** [comment]
@enmasse-ci run tests profile=isolated-shared-infra crc

**Comment by lulf** [comment]
@enmasse-ci run tests profile=isolated-shared-infra

**Comment by lulf** [comment]
@enmasse-ci run tests profile=isolated-shared-infra

**Comment by lulf** [comment]
@enmasse-ci run tests profile=isolated-shared-infra

**Comment by lulf** [comment]
@enmasse-ci run tests profile=isolated-shared-infra

**Comment by lulf** [comment]
@enmasse-ci run tests profile=isolated-shared-infra

**Comment by lulf** [comment]
@enmasse-ci run tests profile=isolated-shared-infra crc

**Comment by lulf** [comment]
@enmasse-ci run tests profile=isolated-shared-infra crc

**Comment by lulf** [comment]
@enmasse-ci run tests profile=isolated-shared-infra

**Comment by lulf** [comment]
@enmasse-ci run tests profile=isolated-shared-infra

**Comment by lulf** [comment]
@enmasse-ci run tests profile=isolated-shared-infra

**Comment by kornys** [comment]
@lulf for WS you don't need only web tests, rhea in external clients can send/receive messages with WS without web browser as well.

**Comment by lulf** [comment]
@kornys Ah, I wasn't aware of that. I will change it to use external clients, thanks!

**Comment by lulf** [comment]
@enmasse-ci run tests profile=isolated-shared-infra

**Comment by lulf** [comment]
@enmasse-ci run tests profile=isolated-shared-infra crc

File: pkg/controller/messaginginfra/cert/controller.go [hunk 1]
```diff
@@ -209,20 +391,40 @@ func (c *CertController) applyCertSecret(secret *corev1.Secret, caSecret *corev1
...
 		}
 		secret.Data = make(map[string][]byte)
-		secret.Data["tls.key"] = key
-		secret.Data["tls.crt"] = cert
-		secret.Data["ca.crt"] = caSecret.Data["tls.crt"]
-		secret.Data["keystore.p12"] = keystore
-		secret.Data["truststore.p12"] = truststore
+		if config.key != "" {
+			secret.Data[config.key] = key
+		}
```
Comment (k-wall) [inline pkg/controller/messaginginfra/cert/controller.go:401]: I know we don't expect the shape of the secret to change, but we avoid the risk of leaving stale values behind if we do this:
```suggestion
		} else {
			   remove(secret.Data[config.key])
		}
```

File: pkg/controller/messaginginfra/cert/controller.go [hunk 2]
```diff
@@ -180,22 +345,39 @@ func (c *CertController) applyCertSecret(secret *corev1.Secret, caSecret *corev1
...
 
 	now := c.clock.Now()
+	var certInfo *CertInfo
 
 	// Create the initial certificate if this is a new secret to be created
 	if !hasCert {
 		logger.Info("Creating component certificate", "secret", secret.Name)
 		expiryDate := now.Add(c.certExpirationTime)
-		key, cert, keystore, truststore, err := generateCert(nil, nil, caSecret.Data["tls.key"], caSecret.Data["tls.crt"], expiryDate, commonName, dnsNames)
+		key, cert, keystore, truststore, err := generateCert(now, nil, nil, caSecret.Data["tls.key"], caSecret.Data["tls.crt"], expiryDate, commonName, dnsNames)
```
Comment (k-wall) [inline pkg/controller/messaginginfra/cert/controller.go:354]: I wonder if we should generate the cert with a valid from 1h in the past, to help guard against clock skew causing spurious connection failures.

**Comment by lulf** [comment]
@enmasse-ci run tests profile=isolated-shared-infra

File: pkg/state/router.go [hunk 3]
```diff
@@ -620,7 +620,7 @@ func entityToMap(v interface{}) (map[string]interface{}, error) {
 	converted := make(map[string]interface{}, len(data))
 
 	for k, v := range data {
-		switch vt := v.(type) {
+		switch v.(type) {
```
Comment (k-wall) [inline pkg/state/router.go:623]: If you retained the vt variable, you could avoid the cast.   That's really why type switches exist.

File: pkg/controller/messaginginfra/cert/controller.go [hunk 4]
```diff
@@ -209,20 +393,42 @@ func (c *CertController) applyCertSecret(secret *corev1.Secret, caSecret *corev1
...
-		secret.Data["keystore.p12"] = keystore
-		secret.Data["truststore.p12"] = truststore
+		if config.key != "" {
+			secret.Data[config.key] = key
+		} else {
+			delete(secret.Data, config.key)
+		}
+		if config.crt != "" {
+			secret.Data[config.crt] = cert
+		}
```
Comment (k-wall) [inline pkg/controller/messaginginfra/cert/controller.go:408]: My intention was that we'd apply the same else {} pattern to the other ifs to.

**Pull request merged**